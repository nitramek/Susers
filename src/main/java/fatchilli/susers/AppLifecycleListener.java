package fatchilli.susers;

public interface AppLifecycleListener {
    void onStart();

    void onFinish();
}
