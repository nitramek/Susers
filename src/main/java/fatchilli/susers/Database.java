package fatchilli.susers;

import java.sql.Connection;
import java.sql.SQLException;

public interface Database {
    Connection createConnection() throws SQLException;
}
