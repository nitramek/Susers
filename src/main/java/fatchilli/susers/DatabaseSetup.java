package fatchilli.susers;

import java.io.IOException;
import java.io.InputStream;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Properties;
import java.util.stream.Collectors;

import javax.sql.DataSource;

import org.hsqldb.jdbc.JDBCDataSourceFactory;

public class DatabaseSetup implements AppLifecycleListener, Database {

    private final DataSource dataSource;

    public DatabaseSetup() {
        final Properties properties = new Properties();
        try (final InputStream is = this.getClass().getResource("/db.properties").openStream()) {
            properties.load(is);
        } catch (IOException e) {
            throw new SystemException("Couldn't find file with db information?", e);
        }
        try {
            //normally datasource would be taken from JNDI or similar mechanism
            dataSource = JDBCDataSourceFactory.createDataSource(properties);
        } catch (Exception e) {
            throw new SystemException("DataSource couldn't be established", e);
        }

    }

    @Override
    public void onStart() {
        try (Connection connection = dataSource.getConnection()) {
            final String sql = Files.lines(Paths.get(this.getClass().getResource("/init.sql").toURI()))
                    .collect(Collectors.joining());
            try (Statement statement = connection.createStatement()) {
                statement.execute(sql);
            }

        } catch (SQLException | URISyntaxException | IOException e) {
            throw new SystemException("Couldn't create database", e);
        }
    }

    @Override
    public void onFinish() {
        //no op
    }

    @Override
    public Connection createConnection() throws SQLException {
        return dataSource.getConnection();
    }
}
