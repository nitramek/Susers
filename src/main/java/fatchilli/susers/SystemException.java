package fatchilli.susers;

/**
 * Marks there is an unpredictable and (most likely) unrecoverable exception
 */
public class SystemException extends RuntimeException {

    public SystemException(String message) {
        super(message);
    }

    public SystemException(String message, Throwable cause) {
        super(message, cause);
    }
}
