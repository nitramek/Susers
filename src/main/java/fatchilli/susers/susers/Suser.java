package fatchilli.susers.susers;

import java.io.Serializable;

import lombok.NonNull;
import lombok.Value;

@Value(staticConstructor = "of")
public class Suser implements Serializable {

    long userId;

    @NonNull
    String guid;
    @NonNull
    String username;

}
