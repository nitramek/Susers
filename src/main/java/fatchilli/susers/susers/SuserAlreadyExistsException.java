package fatchilli.susers.susers;

import lombok.Getter;

public class SuserAlreadyExistsException extends RuntimeException {

    @Getter
    private final long suserId;

    SuserAlreadyExistsException(long suserId) {
        super(String.format("Suser with id %d already exists", suserId));
        this.suserId = suserId;
    }
}
