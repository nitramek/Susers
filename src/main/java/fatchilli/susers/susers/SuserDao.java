package fatchilli.susers.susers;

import java.util.List;

public interface SuserDao {
    void add(Suser suser) throws SuserAlreadyExistsException;

    void deleteSusers();

    List<Suser> getSusers();
}
