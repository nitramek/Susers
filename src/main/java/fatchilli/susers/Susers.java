package fatchilli.susers;

import java.util.List;

import fatchilli.susers.cli.Add;
import fatchilli.susers.cli.DeleteAll;
import fatchilli.susers.cli.PrintAll;
import fatchilli.susers.cli.SusersInputProcessor;
import fatchilli.susers.susers.SuserDaoImpl;

public class Susers {

    public static void main(String[] args) {
        final DatabaseSetup databaseSetup = new DatabaseSetup();

        final SuserDaoImpl dao = new SuserDaoImpl(databaseSetup);
        final List<AppLifecycleListener> databaseSetups = List.of(databaseSetup,
                new SusersInputProcessor(System.in, System.out,
                        List.of(new Add(dao), new PrintAll(dao), new DeleteAll(dao))));

        for (AppLifecycleListener setup : databaseSetups) {
            setup.onStart();
        }

        for (AppLifecycleListener setup : databaseSetups) {
            setup.onFinish();
        }
    }
}
