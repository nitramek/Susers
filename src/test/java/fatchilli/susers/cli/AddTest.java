package fatchilli.susers.cli;

import static org.assertj.core.api.Assertions.assertThat;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import org.approvaltests.Approvals;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

import fatchilli.susers.susers.Suser;
import fatchilli.susers.susers.SuserAlreadyExistsException;
import fatchilli.susers.susers.SuserDao;

class AddTest {

    @Test
    void acceptCorrectLine() throws Exception {
        final Stream<String> inputs = Files.lines(Path.of(this.getClass().getResource("/add.correct.input").toURI()));
        List<Suser> result = new ArrayList<>();
        Assertions.assertThatNoException()
                .isThrownBy(() -> {
                    final SuserDao stub = new SuserDao() {
                        @Override
                        public void add(Suser suser) throws SuserAlreadyExistsException {
                            result.add(suser);
                        }

                        @Override
                        public void deleteSusers() {
                            throw new UnsupportedOperationException();
                        }

                        @Override
                        public List<Suser> getSusers() {
                            throw new UnsupportedOperationException();
                        }
                    };
                    final Add addCommand = new Add(stub);
                    inputs.forEach(input -> {
                        assertThat(addCommand.doesAcceptLine(input)).isTrue()
                                .describedAs("Add command has to accept before processing");
                        addCommand.processLine(input);
                    });

                });
        Approvals.verifyAll("Resulting users", result);
    }

    @Test
    void failsOnIncorrectFormat() throws Exception {
        final Stream<String> inputs = Files.lines(Path.of(this.getClass().getResource("/add.incorrect.input").toURI()));
        final Add add = new Add(new SuserDao() {
            @Override
            public void add(Suser suser) throws SuserAlreadyExistsException {

            }

            @Override
            public void deleteSusers() {

            }

            @Override
            public List<Suser> getSusers() {
                return null;
            }
        });
        inputs.forEach(input -> Assertions.assertThatExceptionOfType(CommandParseException.class)
                .isThrownBy(() -> add.processLine(input)));
    }

    @Test
    void ignoringOtherCommand() {
        final Add add = new Add(new SuserDao() {
            @Override
            public void add(Suser suser) throws SuserAlreadyExistsException {

            }

            @Override
            public void deleteSusers() {

            }

            @Override
            public List<Suser> getSusers() {
                return null;
            }
        });

        assertThat(add.doesAcceptLine("PrintAll"))
                .isFalse();

    }
}