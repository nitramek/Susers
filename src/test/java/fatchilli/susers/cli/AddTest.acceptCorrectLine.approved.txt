Resulting users[0] = Suser(userId=1, guid=a1, username=Robert)
Resulting users[1] = Suser(userId=2, guid=a2, username=Martin)
Resulting users[2] = Suser(userId=4, guid=a2, username=Potato)
Resulting users[3] = Suser(userId=4, guid=, username=)
Resulting users[4] = Suser(userId=5, guid=, username=Empty)
Resulting users[5] = Suser(userId=6, guid='hi', username=Empty)
Resulting users[6] = Suser(userId=6, guid=/ğŸ‘Œ, username=Empty)
